[Fix]: ScriptObject lifting issue

jiashenC opened the pull request:
#### Issue
ScriptObject was treated as normal attribute by the converter previously. This PR lifts it to be a constant and convert it directly to a GetAttr fx node. ScriptObject would also trigger `CallMethod` and this PR adds that support as well.

#### Test Plan
Add test case for ScriptObject.
`pytest test/export/test_converter.py -s -k test_convert_script_object`

angelayi commented on test/export/test_converter.py:
```diff
@@ -966,6 +1010,20 @@ def func2(*args):
+                return torch.ops._TorchScriptTesting.queue_pop(self.tq), self.tq.pop()
+
+        inp = (torch.randn(2, 3),)
+        self._check_equal_ts_ep_converter(M(), inp, ["script"])
```
does jit.trace not support?

jiashenC replied:
Right, both push call does not return anything, so trace does not work.

angelayi commented on torch/_export/converter.py:
```diff
@@ -450,6 +450,20 @@ def convert_prim_Constant(self, node: torch._C.Node):
+        def call_method(inst, name, *args, **kwargs):
+            return getattr(inst, name)(*args, **kwargs)
+
+        fx_node = self.fx_graph.call_function(call_method, (
```
I believe you can use `self.fx_graph.call_method`

angelayi commented on torch/_export/converter.py:
```diff
@@ -988,10 +1026,14 @@ def convert(self) -> ExportedProgram:
+            self.name_to_constant,
         )
         gm = graph_converter.convert()
+        gm.print_readable()
```
delete?

angelayi commented on torch/_export/converter.py:
```diff
@@ -1016,7 +1059,10 @@ def explain(self):
-        self, gm: torch.fx.GraphModule, tensor_constants: Dict[str, torch.Tensor]
+        self,
+        gm: torch.fx.GraphModule,
+        name_to_tensor_constants: Dict[str, torch.Tensor],
```
could we combine name_to_tensor_constants with name_to_constant?

Commit:
[Fix]: ScriptObject lifting issue (#130952)

Summary:
ScriptObject was treated as normal attribute by the converter previously. This PR lifts it to be a constant and convert it directly to a GetAttr fx node. ScriptObject would also trigger `CallMethod` and this PR adds that support as well.

Test Plan:
Add test case for ScriptObject.
`pytest test/export/test_converter.py -s -k test_convert_script_object`

Differential Revision: D60177312

Pulled By: jiashenC

Commit:
fix lint

Commit:
merge name_to_constant/tensor_constant

jiashenC commented on torch/_export/converter.py:
```diff
@@ -465,9 +503,8 @@ def convert_prim_GetAttr(self, node: torch._C.Node):
 
-        attr_value = node.output()
         if self.is_top_level_graph():
-            if attr_value.type().annotation_str == "Tensor":
```
I find this check to be not reliable, e.g., sometimes this is `Optional[Tensor]`.

Commit:
update how to populate constants

Commit:
fix constants

angelayi commented on torch/_export/converter.py:
```diff
@@ -340,8 +350,10 @@ def get_fx_value(self, value: torch._C.Value):
-            return self.constant_map[value_name]
+        elif value_name in self.name_to_constant:
+            if isinstance(self.name_to_constant[value_name], torch.ScriptObject):
+                return self.fx_graph.get_attr(value_name)
```
do we need to manually create the getattr node? Does `return self.constant_map[value_name]` not just automatically create the getattr node?

jiashenC replied:
Right, I think we discussed that we want to only keep the actual value in the `name_to_constant` map. And every time any nodes tries to use the actual value, we insert an ad-hoc `getattr`.

angelayi commented on torch/_export/converter.py:
```diff
@@ -1082,24 +1133,38 @@ def retrace_as_exported_program(
+            if (
+                spec.kind == InputKind.BUFFER
+                and spec.target in name_to_constant
+                and isinstance(name_to_constant[spec.target], torch.Tensor)
```
we should mark ScriptObjects as InputKind.CUSTOM_OBJ

angelayi commented on torch/_export/converter.py:
```diff
@@ -397,9 +409,27 @@ def convert_graph_inputs(self):
+                fx_node = get_node_as_placeholder_or_get_attr(
                     self.fx_graph, name, self.is_top_level_graph()
                 )
+            elif name in self.name_to_constant:
```
Don't you need to check that this input is a ScriptObject?

jiashenC replied:
My reasoning is only ScriptObject will be lifted as an inputs (i.e., showing up in `ts_graph.inputs()`), so we don't really need to check? Correct me if I am wrong.

Commit:
switch to assertion

Commit:
add another assertion for inputs
